src/jit/liveness.cpp

```diff
@@ -1098,34 +1104,47 @@ void Compiler::fgExtendDbgLifetimes()
                 continue;
             }
 
-            // TODO-LIR: the code below does not work for blocks that contain LIR. As a result,
-            //           we must run liveness at least once before any LIR is created in order
-            //           to ensure that this code doesn't attempt to insert HIR into LIR blocks.
-
             // If we haven't already done this ...
             if (!fgLocalVarLivenessDone)
             {
-                assert(!block->IsLIR());
-
                 // Create a "zero" node
-
-                GenTreePtr zero = gtNewZeroConNode(genActualType(type));
+                GenTree* zero = gtNewZeroConNode(genActualType(type));
 
                 // Create initialization node
+                if (!block->IsLIR())
+                {
+                    GenTree* varNode = gtNewLclvNode(varNum, type);
+                    GenTree* initNode = gtNewAssignNode(varNode, zero);
+
+                    // Create a statement for the initializer, sequence it, and append it to the current BB.
+                    GenTree* initStmt = gtNewStmt(initNode);
+                    gtSetStmtInfo(initStmt);
+                    fgSetStmtSeq(initStmt);
+                    fgInsertStmtNearEnd(block, initStmt);
+                }
+                else
+                {
+                    GenTree* store = new (this, GT_STORE_LCL_VAR) GenTreeLclVar(GT_STORE_LCL_VAR, type, varNum, BAD_IL_OFFSET);
+                    store->gtOp.gtOp1 = zero;
+                    store->gtFlags |= (GTF_VAR_DEF | GTF_ASG);
 
-                GenTreePtr varNode  = gtNewLclvNode(varNum, type);
-                GenTreePtr initNode = gtNewAssignNode(varNode, zero);
-                GenTreePtr initStmt = gtNewStmt(initNode);
-
-                gtSetStmtInfo(initStmt);
+                    LIR::Range& blockRange = LIR::AsRange(block);
 
-                /* Assign numbers and next/prev links for this tree */
+                    GenTree* insertionPoint = nullptr;
+                    if ((block->bbJumpKind == BBJ_COND) || (block->bbJumpKind == BBJ_SWITCH) || (block->bbJumpKind == BBJ_RETURN))
```

Comment:
It might be useful to describe why these are special cases. You could even assert what kind of last node you're expecting to see in each case, which would help the reader understand why it's ok to insert a store immediately before that particular node.

---
In reply to: [76294305](https://github.com/dotnet/coreclr/pull/6552#discussion_r76294305) [](ancestors = 76294305)